components/Editor/index.tsx

```diff
@@ -82,49 +93,94 @@ const Editor = ({ readOnly = false }: Props) => {
+        cloned.push({ type, message: line })
+        return cloned
+      })
+    },
+    [setOutput],
+  )
 
-    if (error) {
-      log(error, 'error')
-      setIsCompiling(false)
-      return
-    }
+  const getCallValue = useCallback(() => {
+    const _callValue = new BN(callValue)
 
-    if (warning) {
-      log(warning, 'warn')
+    if (unit === ValueUnit.Gwei) {
+      _callValue.imul(new BN('1000000000'))
+    } else if (unit === ValueUnit.Finney) {
+      _callValue.imul(new BN('1000000000000000'))
+    } else if (unit === ValueUnit.Ether) {
+      _callValue.imul(new BN('1000000000000000000'))
     }
 
-    log('Compilation successful')
+    return _callValue
+  }, [callValue, unit])
 
-    try {
-      deployContract(byteCode, new BN(callValue)).then((tx) => {
-        loadInstructions(byteCode)
+  const handleWorkerMessage = useCallback(
+    (event: MessageEvent) => {
+      const { code: byteCode, warning, error } = event.data
+
+      if (error) {
+        log(error, 'error')
         setIsCompiling(false)
-        startTransaction(byteCode, tx)
-      })
-    } catch (error) {
-      log((error as Error).message, 'error')
-      setIsCompiling(false)
-    }
-  }
+        return
+      }
 
-  const log = useCallback(
-    (line: string, type = 'info') => {
-      output.push({ type, message: line })
-      setOutput(output)
+      if (warning) {
+        log(warning, 'warn')
+      }
+
+      log('Compilation successful')
+
+      try {
+        const _callValue = getCallValue()
+        deployContract(byteCode, _callValue).then((tx) => {
+          loadInstructions(byteCode)
+          setIsCompiling(false)
+          startTransaction(byteCode, tx)
+        })
+      } catch (error) {
+        log((error as Error).message, 'error')
+        setIsCompiling(false)
+      }
     },
-    [output, setOutput],
+    [
+      log,
+      setIsCompiling,
+      deployContract,
+      loadInstructions,
+      startTransaction,
+      getCallValue,
+    ],
   )
 
   useEffect(() => {
-    const initialCodeType: CodeType =
-      getSetting(Setting.EditorCodeType) || CodeType.Yul
+    const query = router.query
 
-    setCodeType(initialCodeType)
-    setCode(examples[initialCodeType][0])
+    if ('callValue' in query && 'unit' in query) {
+      setCallValue(query.callValue as string)
+      setUnit(query.unit as string)
+    }
+
+    if ('callData' in query) {
+      setCallData(query.callData as string)
+    }
+
+    if ('codeType' in query && 'code' in query) {
+      setCodeType(query.codeType as string)
+      setCode(JSON.parse('{"a":' + decode(query.code as string) + '}').a)
+    } else {
+      const initialCodeType: CodeType =
+        getSetting(Setting.EditorCodeType) || CodeType.Yul
+
+      setCodeType(initialCodeType)
+      setCode(examples[initialCodeType][0])
+    }
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [settingsLoaded])
+  }, [settingsLoaded && router.isReady])
```

Comment:
I think these should be comma-separated, so it tracks both separately?

Reply:
I want that function to be called when both the settings are loaded and the router is ready, it's not the way to do it?

Reply:
Ah in that case it's ok. I thought you wanted to run it when any change on deps happen. 👌🏽